File: algorithms/cpp/152.乘积最大子数组.cpp
#include <iostream>
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=152 lang=cpp
 *
 * [152] 乘积最大子数组
 */

// @lc code=start
class Solution {
 public:
  int maxProduct(vector<int>& nums) {
    int len = nums.size();
    vector<int> min_dp(len, 1);
    min_dp[0] = nums[0];
    for (int i = 1; i < len; i++) {
      min_dp[i] =
          min(min_dp[i - 1] * nums[i], min(nums[i], nums[i - 1] * nums[i]));
      nums[i] =
          max(nums[i - 1] * nums[i], max(nums[i], min_dp[i - 1] * nums[i]));
    }
    return *max_element(nums.begin(), nums.end());
  }
};
// @lc code=end

File: algorithms/golang/152.乘积最大子数组.go
/*
 * @lc app=leetcode.cn id=152 lang=golang
 *
 * [152] 乘积最大子数组
 */

// @lc code=start
/**
 * 需要记忆最小乘积和最大乘积
 */
func maxProduct(nums []int) int {
	size := len(nums)
	ret, minRet, maxRet := nums[0], nums[0], nums[0]
	max := func(a, b int) int {
		if a > b {
			return a
		} else {
			return b
		}
	}
	min := func(a, b int) int {
		if a > b {
			return b
		} else {
			return a
		}
	}
	for i := 1; i < size; i++ {
		if nums[i] < 0 {
			minRet, maxRet = maxRet, minRet
		}
		maxRet = max(maxRet*nums[i], nums[i])
		minRet = min(minRet*nums[i], nums[i])
		ret = max(maxRet, ret)
	}
	return ret
}

// @lc code=end
